### ISSUE TITLE
Is there any way to speed up startup?

### ISSUE THREAD

**Issue by JumpIn-Git** [gabm/Satty#132]
Since satty uses gtk, the startup is pretty slow, is there any way to speed it up? Im using hyprland (arch)

**Comment by JumpIn-Git** [comment]
> do you use the ppm image format? that should be faster (see README).
> 
> Apart from that, we might need to run the initial image reading from stdin in async .. you can try to implement that, if you want.

I want to pipe images into satty, do you mean ppm is faster for loading in  satty or saving the file

**Comment by gabm** [comment]
ppm is faster for loading, because it's uncompressed. png is using zip inside which is single threaded and may run a tiny bit slower.

As output format we only support png at this point.

**Comment by JumpIn-Git** [comment]
> ppm is faster for loading, because it's uncompressed. png is using zip inside which is single threaded and may run a tiny bit slower.
> 
> As output format we only support png at this point.

i think it is slower because its also using libadwaita, is there any way to preload the latter/gtk?

**Comment by gabm** [comment]
I did some benchmarks w.r.t. intra-app startup and there the major factor was png decoding.

You claim that most of the startup time is loading of libadwaita. Might be. Can you back your claim?

**Comment by JumpIn-Git** [comment]
> I did some benchmarks w.r.t. intra-app startup and there the major factor was png decoding.
> 
> You claim that most of the startup time is loading of libadwaita. Might be. Can you back your claim?

im just geussing, i mean you need to load gtk and libadwaita

**Comment by gabm** [comment]
yes you do but I'm not sure that this is bottleneck...

i also feel that satty could start up faster, but I'm not sure what is "slow". We need to investigate that first imho

**Comment by jficz** [comment]
> I did some benchmarks w.r.t. intra-app startup and there the major factor was png decoding.

I did a naive "UX perception" test (started the app and counted in my head :) ) and came to the conclusion that `png` vs `ppm` makes imperceptible difference - the startup was pretty much the same in all test cases. The startup was, however, rather slow (~3s first start, consecutive starts within a short period of time ~1.5s), regardless of the format.

**Comment by RobertMueller2** [comment]
I think the speed up on the consecutive starts speaks for a caching effect. This can be observed with `LD_DEBUG=statistics`.

On my machine, it takes 0.6s before the first instruction (println in my case) in main is performed, 0.24s if it's a consecutive start.

Image loading is variable depending on image, obviously. That's e.g. 0.2s for small 600x400 image and 0.6s for a 10000x10000 noise image, both png, haven't compared ppm.

GUI starting takes another maybe 0.2s for me, although it is not that easy to get an exact measurement for that. Maybe need to combine this with timestamped strace and/or wayland debug log, of course both will slow it down.

I'll certainly analyse this further down the road, but it may take a while to get a grip on it.

My sentiment at the moment:
- caching - can be utilised for a minor speedup -- maybe we can figure out how to trigger it without actually opening an image, for session startup
- the rest of "before first line in main" could be further improved by a daemon. I'm not a fan of that idea, but it may help on slower machines. But somebody would have to implement it.
- the GUI part, similar dimension as the "rest of before line" in main, I don't think we'll be able to do much about that
- image size is a real wild card, obviously, and for bigger images it is the longest wait. 
- GUI loading and image size can maybe be parallelised, if we can get the dimensions first. On slower machines, that might make more of a difference. But due to the former point, you might see a spinning wheel somewhere then.

**Comment by RobertMueller2** [comment]
To get a better understanding on this, would you people be able to provide output with `--profile-startup` added with #167?

```
date "+%s,%N" && LD_DEBUG=statistics  satty  --profile-startup <whatever parameters you normally use>
```

I suppose we're interested in both first and consecutive starts.

**Comment by postsolar** [comment]
[satty-startup.txt](https://github.com/user-attachments/files/20099534/satty-startup.txt)

**Comment by MoonBurst** [comment]
[output.txt](https://github.com/user-attachments/files/20100328/output.txt)
Admittedly, I'm not entirely sure what I'm doing with the command. But I ran it both with my normal grim tie ins, and on it's own. So hopefully that helps.

**Comment by gabm** [comment]
@MoonBurst unfortunately that didnt go well :/ In your case, the order of the arguments got mixed up.. pls try this

```
date +%s,%N; grim -t ppm - | LD_DEBUG=statistics satty -f - --profile-startup
```

Expected output looks something like this

```
-  date +%s,%N; grim -t ppm - | LD_DEBUG=statistics satty -f - --profile-startup
1746693607,478306854
    402083:	
    402083:	runtime linker statistics:
    402083:	 total startup time in dynamic loader: 161743 cycles
    402083:	           time needed for relocation: 4966 cycles (3.0%)
    402083:	                number of relocations: 103
    402083:	     number of relocations from cache: 7
    402083:	       number of relative relocations: 0
    402083:	          time needed to load objects: 69651 cycles (43.0%)
    402083:	
    402083:	runtime linker statistics:
    402083:	 total startup time in dynamic loader: 52720439 cycles
    402083:	           time needed for relocation: 35514489 cycles (67.3%)
    402083:	                number of relocations: 27799
    402083:	     number of relocations from cache: 2283
    402083:	       number of relative relocations: 90268
    402083:	          time needed to load objects: 17055684 cycles (32.3%)
startup timestamp was 1746693607.728922019 2025-40-08 11:40:07
   20 ms time elapsed: "configuration loaded"
   21 ms time elapsed: "loaded gl"
   21 ms time elapsed: "loading image"
   61 ms time elapsed: "image loaded, starting gui"
 1002 ms time elapsed: "app init end"
 3489 ms time elapsed: "gui show event"
 4438 ms time elapsed: "main loop idle"
```

> Admittedly, I'm not entirely sure what I'm doing with the command

you are instructing your system and Satty to print information about its startup timings. We can deduce, *where* the startup time is spent and what to fix/optimise to guarantee a faster startup time.

**Comment by braun-steven** [comment]
I was going to run this, but apparently v0.18.0 doesn't work for me? 

```bash
# Downloaded from GitHub release
❯ ./satty --version && ./satty -f image.jpg
satty 0.18.0
Error reading config file: Error reading file: No such file or directory (os error 2)

# Built 0.18.0 from source
❯ ./target/release/satty --version && ./target/release/satty -f image.jpg
Error reading config file: Error reading file: No such file or directory (os error 2)

# Built 0.17.0 from source
❯ ./target/release/satty --version && ./target/release/satty -f image.jpg
satty 0.17.0
# (Satty successfully opens)
```

I'm on Fedora 42

```bash
❯ uname -a
Linux tuxedo 6.14.4-300.fc42.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Apr 25 15:43:38 UTC 2025 x86_64 GNU/Linux
```

**Comment by braun-steven** [comment]
```
❮ date "+%s,%N" && LD_DEBUG=statistics ./satty  --profile-startup --filename ~/image.png
1746698517,033045260
    480362:	
    480362:	runtime linker statistics:
    480362:	  total startup time in dynamic loader: 6567255 cycles
    480362:	            time needed for relocation: 4748377 cycles (72.3%)
    480362:	                 number of relocations: 28487
    480362:	      number of relocations from cache: 2282
    480362:	        number of relative relocations: 11689
    480362:	           time needed to load objects: 1650203 cycles (25.1%)
startup timestamp was 1746698517.061093868 2025-01-08 12:01:57
    0 ms time elapsed: "configuration loaded"
    0 ms time elapsed: "loaded gl"
    0 ms time elapsed: "loading image"
    1 ms time elapsed: "image loaded, starting gui"
  129 ms time elapsed: "app init end"
  242 ms time elapsed: "gui show event"
  319 ms time elapsed: "main loop idle"
    480362:	
    480362:	runtime linker statistics:
    480362:	           final number of relocations: 62934
    480362:	final number of relocations from cache: 16925
```

(default config from README)

**Comment by gabm** [comment]
From reddit: https://www.reddit.com/r/hyprland/comments/1khkoje/comment/mr7oq44/?context=3

```
jtad@archlinux ~ $ date "+%s,%N" && LD_DEBUG=statistics  satty  --profile-startup  --filename img/777.jpeg 
1746693501,031442644
      5385:
      5385:runtime linker statistics:
      5385: total startup time in dynamic loader: 4901935 cycles
      5385:           time needed for relocation: 4033372 cycles (82.2%)
      5385:                number of relocations: 28147
      5385:     number of relocations from cache: 2544
      5385:       number of relative relocations: 20681
      5385:          time needed to load objects: 774219 cycles (15.7%)
startup timestamp was 1746693501.080420836 2025-38-08 10:38:21
    0 ms time elapsed: "configuration loaded"
    1 ms time elapsed: "loaded gl"
    1 ms time elapsed: "loading image"
  271 ms time elapsed: "image loaded, starting gui"
  543 ms time elapsed: "app init end"
MESA-INTEL: warning: ../mesa-25.0.5/src/intel/vulkan/anv_formats.c:802: FINISHME: support YUV colorspace with DRM format modifiers
MESA-INTEL: warning: ../mesa-25.0.5/src/intel/vulkan/anv_formats.c:834: FINISHME: support more multi-planar formats with DRM modifiers
  811 ms time elapsed: "gui show event"
 1213 ms time elapsed: "main loop idle"
      5385:
      5385:runtime linker statistics:
      5385:          final number of relocations: 46415
      5385:final number of relocations from cache: 15940
jtad@archlinux ~ $
```

**Comment by paul-ri** [comment]
Adding `-t ppm` to grim on Sway made a significant difference to the startup time, roughly ~3s shaved off. That's great!

The output of:
```
date "+%s,%N"
grim -t ppm - | LD_DEBUG=statistics satty --profile-startup \
    --copy-command "wl-copy" \
    --early-exit \
    --filename - \
    --fullscreen \
    --initial-tool crop \
    --output-filename "${XDG_SCREENSHOTS_DIR}/satty-%Y_%m_%d-%H:%M:%S.png"
```
Is in: [satty.log](https://github.com/user-attachments/files/20101456/satty.log)

**Comment by quadbyte** [comment]
On my 9950X AMD CPU.

```
❯ date "+%s,%N" && LD_DEBUG=statistics  ./satty  --profile-startup -f $HOME/Pictures/Wallpapers/cache/current.png
1746809760,774293750
      8030:
      8030:     runtime linker statistics:
      8030:       total startup time in dynamic loader: 3539540 cycles
      8030:                 time needed for relocation: 2910227 cycles (82.2%)
      8030:                      number of relocations: 27331
      8030:           number of relocations from cache: 2408
      8030:             number of relative relocations: 9773
      8030:                time needed to load objects: 557142 cycles (15.7%)
startup timestamp was 1746809760.783441049 2025-56-09 12:56:00
    0 ms time elapsed: "configuration loaded"
    0 ms time elapsed: "loaded gl"
    0 ms time elapsed: "loading image"
   31 ms time elapsed: "image loaded, starting gui"
   86 ms time elapsed: "app init end"
  122 ms time elapsed: "gui show event"
  204 ms time elapsed: "main loop idle"
```

**Comment by RobertMueller2** [comment]
> [satty-startup.txt](https://github.com/user-attachments/files/20099534/satty-startup.txt)
> 
> [satty-startup-consecutive.txt](https://github.com/user-attachments/files/20099571/satty-startup-consecutive.txt)

This is a dramatic speedup by the second start. But it does not look like it's related to any library caching -- I don't assume you can reproduce the slow start by running `echo 3 | sudo tee /proc/sys/vm/drop_caches` before the second start, correct?

Is there anything unusual in your wayland session log?

Also, is this something you observe with other gtk (or even gtk4) applications?

**Comment by postsolar** [comment]
> I don't assume you can reproduce the slow start by running `echo 3 | sudo tee /proc/sys/vm/drop_caches` before the second start, correct?

It is slower after this command. Consecutive runs are faster again. First two attachments are running it just after this command, and a consecutive run.

> Is there anything unusual in your wayland session log?

How would I check?

> Also, is this something you observe with other gtk (or even gtk4) applications?

They do often require a couple seconds to start unless they're cached.

[satty.txt](https://github.com/user-attachments/files/20148335/satty.txt)

[satty-consecutive.txt](https://github.com/user-attachments/files/20148336/satty-consecutive.txt)

[strace.out.txt](https://github.com/user-attachments/files/20148343/strace.out.txt)

**Comment by postsolar** [comment]
[strace.out.txt](https://github.com/user-attachments/files/20148635/strace.out.txt)

**Comment by RobertMueller2** [comment]
> [strace.out.txt](https://github.com/user-attachments/files/20148635/strace.out.txt)

Thanks, What does strike me as odd at first glance is what looks to me like a high amount of accesses to icon directories/icon caches.

My reference strace (caches dropped) has 30 occurrences of icon-theme.cache, 53 occurrences of openat matching "icons", 1400 occurrences of newfstatat matching "icons".

Your strace output has 2650, 2650, 13000. Perhaps you just have more icons. I'm not even sure if 2600 ends up being a lot of I/O even if everything hast to be read from disk, so also not sure that is enough to explain it. newfstatat is also the syscall with the biggest occurrence in the strace - but of course that doesn't say anything about timing. Only 4 of the icon-theme.cache actually lead to files, the rest are missing.

One additional thing we could try is to add the -T option to strace, in order to get the actual timing of each of the syscalls. See if any stick out, but by summing these up might also help to see if my theory is correct. So that would be, apologies again, `strace -vfTtt`. And maybe we could compare the ones between the initial and consecutive starts as well.

**Comment by postsolar** [comment]
> Perhaps you just have more icons.

`dust --dereference-links --filecount` run from `/etc/profiles/per-user/alice/share/icons` returns ~10k files. Although I have just 1 icons theme installed (but in 16 color variants) and 1 cursors theme. Is ~10k "a lot"?

> Only 4 of the icon-theme.cache actually lead to files, the rest are missing.

Might be a peculiarity of NixOS, yeah.

[strace.out-first.txt](https://github.com/user-attachments/files/20149659/strace.out-first.txt)
[strace.out-consecutive.txt](https://github.com/user-attachments/files/20149658/strace.out-consecutive.txt)

**Comment by RobertMueller2** [comment]
> `dust --dereference-links --filecount` run from `/etc/profiles/per-user/alice/share/icons` returns ~10k files. Although I have just 1 icons theme installed (but in 16 color variants) and 1 cursors theme. Is ~10k "a lot"?

With all the different sizes I don't think it is a lot.
 
> [strace.out-first.txt](https://github.com/user-attachments/files/20149659/strace.out-first.txt) [strace.out-consecutive.txt](https://github.com/user-attachments/files/20149658/strace.out-consecutive.txt)

I've summed up all the times for syscalls that contained "icons", they amount to 300ms. 280ms in consecutive. So maybe some caching effect, or random, and definitely nowhere near the time we're looking for. So I suppose we can scrap that thought.

There are some longer waits around ppoll and futexes, but that certainly is a bit tricky to interpret due to threading. Not really sure at this point.

**Comment by n0kkster** [comment]
hello everybody, has anyone thought about rewriting satty like a daemon? preload gui once at startup, hide window and then just show it when needed, so on every screenshot it will need only to load image, which is a way faster than cold start

**Comment by RobertMueller2** [comment]
> hello everybody, has anyone thought about rewriting satty like a daemon? preload gui once at startup, hide window and then just show it when needed, so on every screenshot it will need only to load image, which is a way faster than cold start

Yes, it's somewhere above in the discussion. (edit: https://github.com/Satty-org/Satty/issues/132#issuecomment-2852296742)

It'd speed things up, but:
- it'd add considerable complexity with a number of places having to check `if daemon` or similar, up to requiring invalidation of some configs or command line parameters
- even for those people that do wait, the main issue is the first start, with some system caching (not fully investigated yet, see discussion above) making consecutive starts considerably faster (similar effect as a daemon really but perhaps not to the same extent)
- on slow machines, loading bigger images will be slow, regardless of daemon or not
- I don't think we have fully optimised regular startup yet

The 1M$ question is if the speed up is worth the additional complexity. And then of course, somebody would have to implement it.

**Comment by n0kkster** [comment]
@RobertMueller2 

> even for those people that do wait, the main issue is the first start, with some system caching (not fully investigated yet, see discussion above) making consecutive starts considerably faster (similar effect as a daemon really but perhaps not to the same extent)

Cannot agree with that, tested personally on my machine, Intel Ultra 7 155H, 32 RAM, no speed up on consecutive starts. On my machine it takes 60-70 ms to load image and 600 ms to start gtk, so, as gtk does not go anywhere, it is still so slow after any number of runs.

> on slow machines, loading bigger images will be slow, regardless of daemon or not

Probably, but i am not sure there are many machines, that will load png that slow. Anyway it still be faster to only load image, than loading image and gtk.

> The 1M$ question is if the speed up is worth the additional complexity. And then of course, somebody would have to implement it.

I forked satty and implemented this, and I can say that it is 100% worth, speed up about 10x. I have done it for personal usage so it is totally vibecoded, cause I do not know rust and have no experience working with it. Also, sometimes it crashes randomly, had not investigated this yet, cause have no time. You can see my fork, if interested, btw there are not so many changes.

**Comment by RobertMueller2** [comment]
> Cannot agree with that, tested personally on my machine, Intel Ultra 7 155H, 32 RAM, no speed up on consecutive starts. On my machine it takes 60-70 ms to load image and 600 ms to start gtk, so, as gtk does not go anywhere, it is still so slow after any number of runs.

Fair point, I believe that, but this isn't something we've ever seen in the data provided so far.
 
> I forked satty and implemented this, and I can say that it is 100% worth, speed up about 10x. I have done it for personal usage so it is totally vibecoded, cause I do not know rust and have no experience working with it. Also, sometimes it crashes randomly, had not investigated this yet, cause have no time. You can see my fork, if interested, btw there are not so many changes.

Well, for a fully vibe coded change we don't really know that it's complete. So the whole amount of effort or complexity is still not really known. Just one example: with this new mode, I can't start more than one instance as soon as the daemon is running. If that's a use case for me, I can't use the daemon. We'd have to consider what the best user experience is for this. And I could imagine there are more questions just like that, depending on what options/command line parameters you use. I'm just concerned it could make the code a bit messy once we try to address everything in this context.

Also, in order to add this to Satty, somebody would have to understand what the code does and/or reimplement it, and make sure there aren't any IP issues with it. These are the generic concerns we have regarding LLM generated code.

That said, I'm seeing a speedup of 100x with your fork on my machine, which I find quite a convincing case. I can't predict when we find time to implement this ourselves. But now I'm pretty sure we'd consider/review PRs for this, so thanks a lot for your input.

**Comment by n0kkster** [comment]
> Well, for a fully vibe coded change we don't really know that it's complete. So the whole amount of effort or complexity is still not really known. Just one example: with this new mode, I can't start more than one instance as soon as the daemon is running. If that's a use case for me, I can't use the daemon. We'd have to consider what the best user experience is for this. And I could imagine there are more questions just like that, depending on what options/command line parameters you use. I'm just concerned it could make the code a bit messy once we try to address everything in this context.

Yes, totally agree with you, my solution is like a proof of concept that this approach works and fastens loading. But in real production ofc there are lots of use cases, that have to be taken into account. 

> Also, in order to add this to Satty, somebody would have to understand what the code does and/or reimplement it, and make sure there aren't any IP issues with it. These are the generic concerns we have regarding LLM generated code.

Yes, for sure and this is 100% right. Not really a fan of LLM generated code tbh, but sometimes it is suitable for personal usage. Agrees that this “raw” LLM code doesn’t have to be merged into the upstream. 

So, I see your points and agree with them, thanks for your opinion. Maybe somebody,
who facing the same problem may found my solution suitable for personal usage, that’s why I wrote about it here.

**Comment by wongcallum** [comment]
I quite like the idea of a daemon, similar to how the foot terminal has a "client-server" model. I haven't had a deep look into n0kkster's fork yet but I was thinking about a few things we'd have to consider for the implementation:

- How would the client send the image to the daemon? Unix sockets? If so we could probably detect if a socket exists already and refuse to start another daemon if it does
- Does the client or server get to decide the config options (save file path etc.)? Or should we have a default on the server which are overridden by the client
- I'm guessing we'll want to reject multiple connections to the same daemon

Just wanted to put these thoughts out there, I'll see if I have the skill to implement this myself 😅

**Comment by RobertMueller2** [comment]
> I quite like the idea of a daemon, similar to how the foot terminal has a "client-server" model. I haven't had a deep look into n0kkster's fork yet but I was thinking about a few things we'd have to consider for the implementation:

There is an implementation in PR #358, but it's pending some review remarks and needs rebasing. I'm not sure about its current status at this point in time.
 
>     * How would the client send the image to the daemon? Unix sockets? If so we could probably detect if a socket exists already and refuse to start another daemon if it does

Yes, should be Unix sockets so we don't open a new can of worms regarding server security considerations. And it is pretty much available everywhere, including Windows and Mac if we ever try to get it to run there. 

Refusing to start another daemon makes sense if the socket in question is in use. But in order to allow for different scenarios using nested Wayland or X11 sessions, there should be a way to override the socket path by env variable and/or cli switch.

>     * Does the client or server get to decide the config options (save file path etc.)? Or should we have a default on the server which are overridden by the client

The client needs to be able to set all of these. So this leaves client-only or client-overrides-server. `--filename` doesn't make that much sense as a server side setting, so it may be easier to leave everything to the client.

That said, there may be challenges with `--app-id` and css.
 
>     * I'm guessing we'll want to reject multiple connections to the same daemon

No, I think that would be quite limiting. Using more than one Satty instance is a valid use case.

**Comment by n0kkster** [comment]
Wow, somebody really created a PR with this behaviour. That’s great! Hope this will be reviewed and merged someday and we will get official support of this cool feature and enjoy blazingly fast startup :)

**Comment by jficz** [comment]
> > How would the client send the image to the daemon? Unix sockets? If so we could probably detect if a socket exists already and refuse to start another daemon if it does
> Refusing to start another daemon makes sense if the socket in question is in use. 

I would partially disagree here. Depending on how the socket is created, it might remain present if the daemon is killed or crashes. Delegating the socket to, say, systemd, would make satty incompatible with non-systemd systems but could be nicely leveraged on systems with systemd.

IIRC just closing the socket from server side doesn't "release" it, it has to be unlinked (deleted). If one tries to `bind()` to such socket, the bind will fail but doesn't imply whether the server-side process is running or not.

On systemd systems the entire socket management can be delegated but on other systems, other safeguards must be put in place, like a PID file. But a PID file comes with similar limits, especially if we want to allow concurrent daemons. Now the question is whether this is worth exploring or whether it can be delegated to the user - such as through a reasonable error message (i.e. "Socket exists, please check if another Satty instance is running and delete the socket").

I would be for an iterative approach - start with the error message requiring user intervention and see how it goes. Maybe even go with decoupled systemd approach first, leaving the socket logic completely out of Satty (systemd basically just connects a socket to process' stdio).

**Comment by RobertMueller2** [comment]
> > > How would the client send the image to the daemon? Unix sockets? If so we could probably detect if a socket exists already and refuse to start another daemon if it does
> > Refusing to start another daemon makes sense if the socket in question is in use. 
> 
> I would partially disagree here. Depending on how the socket is created, it might remain present if the daemon is killed or crashes. Delegating the socket to, say, systemd, would make satty incompatible with non-systemd systems but could be nicely leveraged on systems with systemd.

Not the point i was trying to make. I merely wanted to say there shouldn't be a hard coded socket path.

**Comment by jficz** [comment]
> Not the point i was trying to make. I merely wanted to say there shouldn't be a hard coded socket path.

I see. That is a good point either way. No paths should ever be hardcoded anywhere :)

**Comment by wongcallum** [comment]
Thanks, I didn't see that PR. I'm glad that this feature is getting attention.

**Comment by RobertMueller2** [comment]
In #685 we were made aware that an NVidia issue wakes up the GPU for Vulkan device enumeration performed by GTK. If the card is suspended, this can considerably increase startup times. For those that have exceessive start types AND an NVidia dGPU, you might want to check that and try starting with `VK_LOADER_DRIVERS_DISABLE=nvidia_icd.json`.

Also described in https://github.com/Satty-org/Satty/wiki/(anticipated)-FAQ#nvidia-dgpu-wakes-up-from-suspend-when-starting-satty

**Comment by the-can-of-soup** [comment]
> In [#685](https://github.com/Satty-org/Satty/issues/685) we were made aware that an NVidia issue wakes up the GPU for Vulkan device enumeration performed by GTK. If the card is suspended, this can considerably increase startup times. For those that have exceessive start types AND an NVidia dGPU, you might want to check that and try starting with `VK_LOADER_DRIVERS_DISABLE=nvidia_icd.json`.
> 
> Also described in https://github.com/Satty-org/Satty/wiki/(anticipated)-FAQ#nvidia-dgpu-wakes-up-from-suspend-when-starting-satty

I fall into this demographic, but it barely helped for me, if at all.

**Comment by RobertMueller2** [comment]
> > In [#685](https://github.com/Satty-org/Satty/issues/685) we were made aware that an NVidia issue wakes up the GPU for Vulkan device enumeration performed by GTK. If the card is suspended, this can considerably increase startup times. For those that have exceessive start types AND an NVidia dGPU, you might want to check that and try starting with `VK_LOADER_DRIVERS_DISABLE=nvidia_icd.json`.
> > Also described in https://github.com/Satty-org/Satty/wiki/(anticipated)-FAQ#nvidia-dgpu-wakes-up-from-suspend-when-starting-satty
> 
> I fall into this demographic, but it barely helped for me, if at all.

the main question for this is, is the dGPU sleeping at the time of starting Satty. If it is not, then this isn't making a difference. That's the situation I'm personally observing on my system, it is an AMD/NVidia laptop, but my PM config is broken and the dGPU is awake already. Not enumerating it does save some time, but it's not much.

What is your start time according to `--profile-startup`?

**Comment by the-can-of-soup** [comment]
> > > ...
> > 
> > 
> > I fall into this demographic, but it barely helped for me, if at all.
> 
> the main question for this is, is the dGPU sleeping at the time of starting Satty. If it is not, then this isn't making a difference. That's the situation I'm personally observing on my system, it is an AMD/NVidia laptop, but my PM config is broken and the dGPU is awake already. Not enumerating it does save some time, but it's not much.
> 
> What is your start time according to `--profile-startup`?

```
   63 ms time elapsed: image loaded, starting gui
  119 ms time elapsed: app init end
 2409 ms time elapsed: gui show event
 2506 ms time elapsed: main loop idle
```

With `VK_LOADER_DRIVERS_DISABLE=nvidia_icd.json`:
```
   94 ms time elapsed: image loaded, starting gui
  137 ms time elapsed: app init end
 2420 ms time elapsed: gui show event
 2505 ms time elapsed: main loop idle
```

**Comment by RobertMueller2** [comment]
> ```
>    63 ms time elapsed: image loaded, starting gui
>   119 ms time elapsed: app init end
>  2409 ms time elapsed: gui show event
>  2506 ms time elapsed: main loop idle
> ```
> 
> With `VK_LOADER_DRIVERS_DISABLE=nvidia_icd.json`:
> 
> ```
>    94 ms time elapsed: image loaded, starting gui
>   137 ms time elapsed: app init end
>  2420 ms time elapsed: gui show event
>  2505 ms time elapsed: main loop idle
> ```

that's definitely long, probably worth recording `strace -f -tt -T` to see if anything sticks out.
